### PR TITLE
API change to improve VolumeGroupSnapshot restore

### PR DESCRIPTION
**Pull request by leonardoce**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR implements an API change proposed by @xing-yang (see https://docs.google.com/document/d/1NdNwFD5Z64K2heQLYOnojt6Ogulg750BuYIJ6W9cEiM/edit?usp=sharing and the relative conversation in #969).

To ease the review process, this PR has been split into two commits:

1. a commit changing the API, including the generated files
2. some minimal changes on the controller and the CSI sidecar to allow the code to compile with the new API.

The information we have in `VolumeGroupSnapshot.status.volumeSnapshotRefList` has been moved into `VolumeGroupSnapshot.status.pvcVolumeSnapshotRefList.volumeSnapshotRef`.

The information we have in `VolumeGroupSnapshotContent.status.volumeSnapshotContentRefList` has been moved into `VolumeGroupSnapshotContent.status.volumeSnapshotContentRefList.volumeSnapshotContentName`.

The code to set the new fields has already been implemented, and I'm going to open a new PR based on this one.
I hope that it will be helpful for the reviewer.

If you prefer a single PR on multiple commits, just tell me.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This is my first API-change PR. I followed the process described in `client/hack/README.md` as carefully as possible, but I encourage you to use special care when reviewing this.

**Testing**

After this patch, a `VolumeGroupSnapshot` will be defined as:

```
apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
kind: VolumeGroupSnapshot
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"groupsnapshot.storage.k8s.io/v1alpha1","kind":"VolumeGroupSnapshot","metadata":{"annotations":{},"name":"new-groupsnapshot-demo","namespace":"default"},"spec":{"source":{"selector":{"matchLabels":{"cnpg.io/cluster":"cluster-example"}}},"volumeGroupSnapshotClassName":"csi-hostpath-groupsnapclass"}}
  creationTimestamp: "2024-04-24T21:09:22Z"
  finalizers:
  - groupsnapshot.storage.kubernetes.io/volumegroupsnapshot-bound-protection
  generation: 1
  name: new-groupsnapshot-demo
  namespace: default
  resourceVersion: "664981"
  uid: e647e38f-d73e-4061-9113-c999d6ebd173
spec:
  source:
    selector:
      matchLabels:
        cnpg.io/cluster: cluster-example
  volumeGroupSnapshotClassName: csi-hostpath-groupsnapclass
status:
  boundVolumeGroupSnapshotContentName: groupsnapcontent-e647e38f-d73e-4061-9113-c999d6ebd173
  creationTime: "2024-04-24T21:09:23Z"
  pvcVolumeSnapshotRefList:
  - persistentVolumeClaimRef: {}
    volumeSnapshotRef:
      name: snapshot-527c30405cb932a46be1207bec23311ded70d6b0ffbfbe534680b8ebd24faea1-2024-04-24-9.9.24
  - persistentVolumeClaimRef: {}
    volumeSnapshotRef:
      name: snapshot-d7734103b40865afaa62b7aa69165c751ae07e9d3a153c62da33bdfe997e1170-2024-04-24-9.9.24
  readyToUse: true
```

And the corresponding VolumeGroupSnapshotContent would be like:

```
apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
kind: VolumeGroupSnapshotContent
metadata:
  creationTimestamp: "2024-04-24T21:09:22Z"
  finalizers:
  - groupsnapshot.storage.kubernetes.io/volumegroupsnapshotcontent-bound-protection
  generation: 1
  name: groupsnapcontent-e647e38f-d73e-4061-9113-c999d6ebd173
  resourceVersion: "664972"
  uid: f7eab355-63c6-4f9d-82a7-ef6768e74a17
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    volumeHandles:
    - d9c0e0b9-020d-11ef-962a-0abf9c1ed13e
    - d9c87ebe-020d-11ef-962a-0abf9c1ed13e
  volumeGroupSnapshotClassName: csi-hostpath-groupsnapclass
  volumeGroupSnapshotRef:
    apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
    kind: VolumeGroupSnapshot
    name: new-groupsnapshot-demo
    namespace: default
    resourceVersion: "664940"
    uid: e647e38f-d73e-4061-9113-c999d6ebd173
status:
  creationTime: 1713992963010846995
  pvVolumeSnapshotContentList:
  - persistentVolumeRef: {}
    volumeSnapshotContentRef:
      name: snapcontent-527c30405cb932a46be1207bec23311ded70d6b0ffbfbe534680b8ebd24faea1-2024-04-24-9.9.24
  - persistentVolumeRef: {}
    volumeSnapshotContentRef:
      name: snapcontent-d7734103b40865afaa62b7aa69165c751ae07e9d3a153c62da33bdfe997e1170-2024-04-24-9.9.24
  readyToUse: true
  volumeGroupSnapshotHandle: ec95172d-027e-11ef-9db6-961b7658d99d
```

**Does this PR introduce a user-facing change?**:

```release-note
Update API for group snapshots, easing the restore process.
```

### CODE REVIEW

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 1]
```diff
@@ -103,11 +103,20 @@ type VolumeGroupSnapshotStatus struct {
...
 	Error *snapshotv1.VolumeSnapshotError `json:"error,omitempty" protobuf:"bytes,4,opt,name=error,casttype=VolumeSnapshotError"`
 
-	// VolumeSnapshotRefList is the list of volume snapshot references for this
-	// group snapshot.
+	// VolumeSnapshotRefList is the list of PVC and VolumeSnapshot pair that
+	// is part of this group snapshot.
 	// The maximum number of allowed snapshots in the group is 100.
 	// +optional
-	VolumeSnapshotRefList []core_v1.ObjectReference `json:"volumeSnapshotRefList,omitempty" protobuf:"bytes,5,opt,name=volumeSnapshotRefList"`
+	PVCVolumeSnapshotRefList []PVCVolumeSnapshotPair `json:"pvcVolumeSnapshotRefList,omitempty" protobuf:"bytes,6,opt,name=pvcVolumeSnapshotRefList"`
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:110]: This should still be 5, not 6.

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 2]
```diff
@@ -103,11 +103,20 @@ type VolumeGroupSnapshotStatus struct {
...
+}
+
+// PVCVolumeSnapshotPair defines a pair of a PVC reference and a Volume Snapshot
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:113]: Volume Snapshot reference

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 3]
```diff
@@ -103,11 +103,20 @@ type VolumeGroupSnapshotStatus struct {
...
+type PVCVolumeSnapshotPair struct {
+	// PersistentVolumeClaimRef is a reference to the PVC this pair is referring to
+	PersistentVolumeClaimRef core_v1.ObjectReference `json:"persistentVolumeClaimRef,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimRef"`
+
+	// VolumeSnapshotRef is a reference to the VolumeSnapshot this pair is referring to
+	VolumeSnapshotRef core_v1.ObjectReference `json:"volumeSnapshotRef,omitempty" protobuf:"bytes,1,opt,name=volumeSnapshotRef"`
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:119]: 2

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 4]
```diff
@@ -337,11 +346,21 @@ type VolumeGroupSnapshotContentStatus struct {
 	// +optional
 	Error *snapshotv1.VolumeSnapshotError `json:"error,omitempty" protobuf:"bytes,4,opt,name=error,casttype=VolumeSnapshotError"`
 
-	// VolumeSnapshotContentRefList is the list of volume snapshot content references
-	// for this group snapshot.
+	// PVVolumeSnapshotContentRefList is the list of pairs of PV and
+	// VolumeSnapshotCOntent for this group snapshot
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:350]: PVVolumeSnapshotContentRefList -> PVVolumeSnapshotContentList
VolumeSnapshotCOntent -> VolumeSnapshotContent names

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 5]
```diff
@@ -337,11 +346,21 @@ type VolumeGroupSnapshotContentStatus struct {
...
 	// The maximum number of allowed snapshots in the group is 100.
 	// +optional
-	VolumeSnapshotContentRefList []core_v1.ObjectReference `json:"volumeSnapshotContentRefList,omitempty" protobuf:"bytes,5,opt,name=volumeSnapshotContentRefList"`
+	PVVolumeSnapshotContentRefList []PVVolumeSnapshotContentPair `json:"pvVolumeSnapshotContentRefList,omitempty" protobuf:"bytes,6,opt,name=pvVolumeSnapshotContentRefList"`
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:353]: PVVolumeSnapshotContentRefList -> PVVolumeSnapshotContentList

File: client/apis/volumegroupsnapshot/v1alpha1/types.go [hunk 6]
```diff
@@ -337,11 +346,21 @@ type VolumeGroupSnapshotContentStatus struct {
...
+}
+
+// PVVolumeSnapshotContentPair represent a pair of PV names and
+// VolumeSnapshotContent names
+type PVVolumeSnapshotContentPair struct {
+	// PersistentVolumeName is the name of the persistent volume
+	PersistentVolumeName string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
+
+	// VolumeSnapshotContentName is the name of the volume snapshot content resource
+	VolumeSnapshotContentName string `json:"volumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=volumeSnapshotContentName"`
```
Comment (xing-yang) [inline client/apis/volumegroupsnapshot/v1alpha1/types.go:363]: 2

**Comment by leonardoce** [comment]
Thank you @xing-yang for reviewing this. I should have taken care of your comments.

File: client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml [hunk 7]
```diff
@@ -183,79 +183,101 @@ spec:
                     format: date-time
                     type: string
                 type: object
+              pvcVolumeSnapshotRefList:
+                description: VolumeSnapshotRefList is the list of PVC and VolumeSnapshot
+                  pair that is part of this group snapshot. The maximum number of
```
Comment (bswartz) [inline client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml:188]: pairs that are part

File: client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml [hunk 8]
```diff
@@ -183,79 +183,101 @@ spec:
...
+                            (where "name" refers to the name of the container that
+                            triggered the event) or if no container name is specified
+                            "spec.containers[2]" (container with index 2 in this pod).
+                            This syntax is chosen only to have some well-defined way
+                            of referencing a part of an object. TODO: this design
+                            is not final and this field is subject to change in the
+                            future.'
+                          type: string
+                        kind:
+                          description: 'Kind of the referent. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
```
Comment (bswartz) [inline client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml:216]: Kind must be PVC!

File: client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml [hunk 9]
```diff
@@ -183,79 +183,101 @@ spec:
...
```
Comment (bswartz) [inline client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshots.yaml:255]: Kind must be VolumeSnapshot!

**Review by bswartz — CHANGES_REQUESTED** [review]
Actually after reviewing the change, I'm wondering why use object refs at all, if the PVC ref will always be a PVC and the snapshot ref will always be a VolumeSnapshot. Can't we do like the non-namespaced objects do and specify the object names, but omit their kinds? Namespace can be implied by the namespace of the containing object.

**Comment by xing-yang** [comment]
> Actually after reviewing the change, I'm wondering why use object refs at all, if the PVC ref will always be a PVC and the snapshot ref will always be a VolumeSnapshot. Can't we do like the non-namespaced objects do and specify the object names, but omit their kinds? Namespace can be implied by the namespace of the containing object.

"Name" might be sufficient for now.  I'm wondering we may want to keep the namespace just in case in the future there's namespace transfer support.

**Comment by leonardoce** [comment]
I replaced ObjectReferences with LocalObjectReferences and updated the code correspondingly. Thank you @bswartz @xing-yang

**Comment by xing-yang** [comment]
LocalObjectReference looks good to me. We should probably use the same construct for PV and VolumeSnapshotContent pairs.
https://github.com/kubernetes/kubernetes/blob/v1.31.0-alpha.0/pkg/apis/core/types.go#L5460
cc @bswartz

**Comment by leonardoce** [comment]
> LocalObjectReference looks good to me. We should probably use the same construct for PV and VolumeSnapshotContent pairs.

Nice. I changed the code for this and updated the example in the PR description.

**Comment by bswartz** [comment]
> LocalObjectReference looks good to me. We should probably use the same construct for PV and VolumeSnapshotContent pairs. https://github.com/kubernetes/kubernetes/blob/v1.31.0-alpha.0/pkg/apis/core/types.go#L5460 cc @bswartz

Non-namespaced objects definitely should not be using namespaces. Assuming the kind is constant, why not just use an object name?

**Comment by xing-yang** [comment]
>> LocalObjectReference looks good to me. We should probably use the same construct for PV and VolumeSnapshotContent pairs. https://github.com/kubernetes/kubernetes/blob/v1.31.0-alpha.0/pkg/apis/core/types.go#L5460 cc @bswartz

> Non-namespaced objects definitely should not be using namespaces. Assuming the kind is constant, why not just use an object name?

@msau42 From API design point of view,  are there pros and cons of using LocalObjectReference vs just a string here?

**Review by bswartz — APPROVED** [review]
/lgtm

**Pull request merged**